### PR TITLE
fix: prefer the task avoidance API, see #286

### PR DESCRIPTION
**Pull request by DanySK**
@aSemy please check this out.

### CODE REVIEW

File: src/main/kotlin/org.danilopianini.gradle.mavencentral/PublishOnCentral.kt [hunk 1]
```diff
@@ -71,20 +71,20 @@ class PublishOnCentral : Plugin<Project> {
                 project.configureRepository(extension.mavenCentral)
             }
```
Comment (aSemy) [inline src/main/kotlin/org.danilopianini.gradle.mavencentral/PublishOnCentral.kt:72]: I'm not 100% sure what this code is doing, but `.afterEvaluate { }` is a bit risky - it turns into a race-to-the-bottom as all the plugins try to be the last to evaluate. It's the kind of thing that can be replaced by using lazy config. `afterEvaluate { }`

**Comment by aSemy** [comment]
```kotlin
private inline fun <reified T : Task> Project.registerTaskIfNeeded(name: String): Task =
  tasks.findByName(name) ?: project.tasks.register(name, T::class.java).get()
```

using `.get()` breaks task-config-avoidance, as it eagerly creates the task

Try this instead, but I'm not 100% sure if it will work
```kotlin
  private inline fun <reified T : Task> Project.registerTaskIfNeeded(name: String): Provider<Task> =
    tasks.named(name).orElse(project.tasks.register<T>(name))
```

**Comment by aSemy** [comment]
Something else odd that caught my eye. It looks like the package is a single directory named `org.danilopianini.gradle.mavencentral`, not nested directories `org/danilopianini/gradle/mavencentral` 

https://github.com/DanySK/publish-on-central/tree/master/src/main/kotlin/org.danilopianini.gradle.mavencentral

Is that intentional? :)

**Comment by aSemy** [comment]
https://github.com/DanySK/publish-on-central/blob/f5384d28ae90e2847f79650a649337df011bca7f/src/main/kotlin/org.danilopianini.gradle.mavencentral/PublishOnCentral.kt#L57-L58

These two lines can be done with `.configureEach { }`


```kotlin
project.components.configureEach {component ->

  // code from `fun createPublications(component: SoftwareComponent)`

  project.logger.debug("Reacting to the creation of component ${component.name}")
  publications { publications ->
      val name = "${component.name}$publicationName"
      if (publications.none { it.name == name }) {
          publications.create(name, MavenPublication::class) { publication ->
              publication.from(component)
          }
          project.logger.debug("Created new publication $name")
      }
  }
}
```

File: src/main/kotlin/org.danilopianini.gradle.mavencentral/PublishOnCentral.kt [hunk 2]
```diff
@@ -71,20 +71,20 @@ class PublishOnCentral : Plugin<Project> {
...
         }
-        project.plugins.withType(JavaPlugin::class.java) { _ ->
-            project.tasks.withType(JavadocJar::class.java) { javadocJar ->
+        project.plugins.withType(JavaPlugin::class.java).configureEach { _ ->
```
[the remark this answers is not in the document]
Comment (DanySK) [reply src/main/kotlin/org.danilopianini.gradle.mavencentral/PublishOnCentral.kt]: Right, fixed.

File: src/main/kotlin/org.danilopianini.gradle.mavencentral/PublishOnCentral.kt (hunk 1 above)
Comment (aSemy) [reply src/main/kotlin/org.danilopianini.gradle.mavencentral/PublishOnCentral.kt:71]: Looks good! 

I don't think I have a 'Resolve conversation' button to click? I can't see it

image


> Are there other (better) ways to tell Gradle "run this thing only after all properties have been set"?

Using task-config avoidance :) Sometimes `afterEvaluate { }` will do the job though.

**Pull request merged**